simulators/ethereum/engine/clmock.go

```diff
@@ -0,0 +1,365 @@
+	environment as expected, and not as previously modified by another test.
+
+	The test cases can request a lock to "wake up" at a specific point during
+	the PoS block production procedure and, when the CL mocker has reached it,
+	the lock is released to let only one test case do its testing.
+	*/
+	PayloadBuildMutex                SetResetLock
+	NewGetPayloadMutex               SetResetLock
+	NewExecutePayloadMutex           SetResetLock
+	NewHeadBlockForkchoiceMutex      SetResetLock
+	NewSafeBlockForkchoiceMutex      SetResetLock
+	NewFinalizedBlockForkchoiceMutex SetResetLock
+}
+
+func NewCLMocker() *CLMocker {
+	// Init random seed for different purposes
+	rand.Seed(time.Now().Unix())
+
+	// Create the new CL mocker
+	newCLMocker := &CLMocker{
+		EngineClients:               make([]*EngineClient, 0),
+		RandomHistory:               map[uint64]common.Hash{},
+		ExecutedPayloadHistory:      map[uint64]catalyst.ExecutableDataV1{},
+		LatestFinalizedHeader:       nil,
+		PoSBlockProductionActivated: false,
+		BlockProductionMustStop:     false,
+		FirstPoSBlockNumber:         nil,
+		LatestFinalizedNumber:       nil,
+		TTDReached:                  false,
+		NextFeeRecipient:            common.Address{},
+		LatestForkchoice: catalyst.ForkchoiceStateV1{
+			HeadBlockHash:      common.Hash{},
+			SafeBlockHash:      common.Hash{},
+			FinalizedBlockHash: common.Hash{},
+		},
+	}
+
+	// Lock the mutexes. To be unlocked on next PoS block production.
+	newCLMocker.PayloadBuildMutex.LockReset()
+	newCLMocker.NewGetPayloadMutex.LockReset()
+	newCLMocker.NewExecutePayloadMutex.LockReset()
+	newCLMocker.NewHeadBlockForkchoiceMutex.LockReset()
+	newCLMocker.NewSafeBlockForkchoiceMutex.LockReset()
+	newCLMocker.NewFinalizedBlockForkchoiceMutex.LockReset()
+
+	// Start timer to check when the TTD has been reached
+	time.AfterFunc(tTDCheckPeriod, newCLMocker.checkTTD)
+
+	return newCLMocker
+}
+
+// Add a Client to be kept in sync with the latest payloads
+func (cl *CLMocker) AddEngineClient(newEngineClient *EngineClient) {
+	cl.EngineClients = append(cl.EngineClients, newEngineClient)
+}
+
+// Remove a Client to stop sending latest payloads
+func (cl *CLMocker) RemoveEngineClient(removeEngineClient *EngineClient) {
+	i := -1
+	for j := 0; j < len(cl.EngineClients); j++ {
+		if cl.EngineClients[j] == removeEngineClient {
+			i = j
+			break
+		}
+	}
+	if i >= 0 {
+		cl.EngineClients[i] = cl.EngineClients[len(cl.EngineClients)-1]
+		cl.EngineClients = cl.EngineClients[:len(cl.EngineClients)-1]
+	}
+}
+
+// Check whether we have reached TTD and then enable PoS block production.
+// This function must NOT be executed after we have reached TTD.
+func (cl *CLMocker) checkTTD() {
+	if len(cl.EngineClients) == 0 {
+		// We have no clients running yet, we have not reached TTD
+		time.AfterFunc(tTDCheckPeriod, cl.checkTTD)
+		return
+	}
+
+	// Pick a random client to get the total difficulty of its head
+	ec := cl.EngineClients[rand.Intn(len(cl.EngineClients))]
```

Comment:
What is the purpose of choosing a random client?

Reply:
The idea was to cover testing of a client receiving an ExecutionPayload both created by itself and created by a different client.

Originally I thought of running more than 2 clients at the same time and it made sense to have a client picked randomly here, but perhaps we could simply round-robin.